# Retail Decision Intelligence Platform

## Module 1 – Data Understanding

### Objective

The objective of this notebook is to understand the structure, quality, and characteristics of the retail transaction dataset before beginning any preprocessing or analysis.

Understanding the dataset is an essential first step in the analytics lifecycle because it helps identify potential data quality issues and ensures that future analyses are accurate and reliable.

---

## Business Context

The client is an online retail company that wants to improve decision-making using transactional data.

As a Data Analyst, our responsibility is to understand the available data, identify potential issues, and determine whether the dataset is suitable for answering the client's business questions.

---

## Learning Outcomes

By the end of this notebook, we will:

* Load the dataset into Python
* Understand every column
* Identify data types
* Inspect the dataset structure
* Evaluate missing values
* Detect duplicate records
* Perform an initial data quality assessment


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

#Display all columns
pd.set_option("display.max_columns",None)

#Increase display width
pd.set_option("display.width", 1000)

Matplotlib is building the font cache; this may take a moment.


In [4]:
df = pd.read_excel("../data/raw/online_retail.xlsx")

In [5]:
df.head()


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


## Observations from Initial Data Inspection

The dataset was loaded successfully, and the first five records were examined using the `head()` function.

### Key Observations

- Each row represents a single product purchased within an invoice.
- Multiple rows can belong to the same invoice because one customer may purchase multiple products in a single transaction.
- The dataset contains product information, customer information, pricing, quantity, purchase date, and country.
- The available attributes are sufficient for performing sales analysis, customer segmentation, market basket analysis, and sales forecasting.

### Business Understanding

The presence of repeated invoice numbers indicates that a single order can contain multiple products. This transactional structure makes the dataset suitable for identifying purchasing patterns and cross-selling opportunities.

# Dataset Structure

Before cleaning the dataset, it is essential to understand its dimensions, column names, data types, and memory usage. This helps identify potential issues such as incorrect data types or missing values.

In [6]:
#Dataset Dimensions

rows, columns = df.shape
print(f"Number of Rows: {rows:,}")
print(f"Number of Columns: {columns:}")

Number of Rows: 525,461
Number of Columns: 8


In [7]:
#Display Columns Names

print("Columns in the dataset: \n")
for column in df.columns:
    print(column)

Columns in the dataset: 

Invoice
StockCode
Description
Quantity
InvoiceDate
Price
Customer ID
Country


In [8]:
#Dataset Information

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[us]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 32.1+ MB


# Dataset Structure Analysis

The `info()` function provides a high-level summary of the dataset. It displays the number of records, column names, data types, missing values, and memory usage.

Understanding the dataset structure is essential before performing any preprocessing because it helps identify data quality issues and determines the appropriate cleaning strategy.

---

## Observations

### Dataset Size
- Total Records: 525,461
- Total Features: 8

The dataset is sufficiently large to perform meaningful business analysis and resembles a real-world transactional retail dataset.

### Data Types

The dataset contains four different data types:

- Object (Invoice, StockCode, Description, Country)
- Integer (Quantity)
- Float (Price, Customer ID)
- Datetime (InvoiceDate)

The InvoiceDate column has already been recognized as a datetime object, which is ideal for time-series analysis and forecasting.

### Missing Values

Two columns contain missing values:

- Description
- Customer ID

The remaining columns are complete.

Customer ID contains a significant number of missing values. Since customer segmentation depends on Customer ID, this issue will require careful handling during the data cleaning phase.

### Memory Usage

The dataset occupies approximately 32 MB of memory, which can be processed comfortably on a standard laptop without requiring distributed computing tools.

In [9]:
#Missing Values
df.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [ ]:
#Missing values percentage
(df.isnull().sum() / len(df) * 100).round(2)

Invoice         0.00
StockCode       0.00
Description     0.56
Quantity        0.00
InvoiceDate     0.00
Price           0.00
Customer ID    20.54
Country         0.00
dtype: float64

In [12]:
#Duplicate records
df.duplicated().sum()

np.int64(6865)

In [13]:
#Number of unique values in each column
df.nunique()

Invoice        28816
StockCode       4632
Description     4681
Quantity         825
InvoiceDate    25296
Price           1606
Customer ID     4383
Country           40
dtype: int64

# Data Quality Assessment

## Objective

The purpose of this section is to evaluate the completeness, consistency, and uniqueness of the dataset before performing any cleaning operations.

Data quality directly affects the reliability of business insights. Therefore, missing values, duplicate records, and inconsistent entries must be identified before analysis.

---

## Why Data Quality Assessment Matters

Poor-quality data can lead to:

* Incorrect revenue calculations
* Misleading customer segmentation
* Inaccurate sales forecasts
* Wrong business recommendations

A systematic assessment helps determine which issues require cleaning and which represent valid business events such as cancelled orders.

## Observations – Missing Values

* The dataset is largely complete, with only two columns containing missing values.
* `Description` has 2,928 missing values (0.56%), which represents a very small proportion of the dataset.
* `Customer ID` has 107,927 missing values (20.54%), indicating that a substantial number of transactions cannot be linked to individual customers.
* Sales-related analyses can still use these transactions, but customer-level analyses such as RFM segmentation will require valid Customer IDs.

## Observations – Duplicate Records

* The dataset contains **6,865 duplicate rows**, representing approximately **1.31%** of all records.
* Exact duplicates are unlikely to represent additional customer purchases because every attribute is identical.
* These records will be removed during the data cleaning phase to prevent double-counting of revenue and transaction volume.

## Observations – Dataset Diversity

* The dataset contains **28,816 invoices**, indicating a large number of customer transactions.
* There are **4,632 unique products**, providing a broad product portfolio for sales and inventory analysis.
* The dataset includes **4,383 unique customers**, which is sufficient for customer segmentation and retention analysis.
* Customers belong to **40 different countries**, enabling international market performance analysis.
